spec/unit/util/package/version/pip_spec.rb

```diff
@@ -0,0 +1,439 @@
+        it_should_behave_like 'a valid version', "1.0+local"
+        it_should_behave_like 'a valid version', "1.0+Local", "1.0+local"
+      end
+
+      context "when local version contains numbers" do
+        it_should_behave_like 'a valid version', "1.0+10"
+        it_should_behave_like 'a valid version', "1.0+01",    "1.0+1"
+        it_should_behave_like 'a valid version', "1.0+01L",   "1.0+01l"
+        it_should_behave_like 'a valid version', "1.0+L101L", "1.0+l101l"
+      end
+
+      context "when local version contains multiple elements" do
+        it_should_behave_like 'a valid version', "1.0+10.local"
+        it_should_behave_like 'a valid version', "1.0+abc.def.ghi"
+        it_should_behave_like 'a valid version', "1.0+01.abc",            "1.0+1.abc"
+        it_should_behave_like 'a valid version', "1.0+01L.0001",          "1.0+01l.1"
+        it_should_behave_like 'a valid version', "1.0+L101L.local",       "1.0+l101l.local"
+        it_should_behave_like 'a valid version', "1.0+dash-undrsc_dot.5", "1.0+dash.undrsc.dot.5"
+      end
+    end
+  end
+
+  describe "comparison of versions" do
+    # This array must remain sorted (smallest to highest version).
+    versions = [
+      "0.1",
+      "0.10",
+      "0.10.1",
+      "0.10.1.0.1",
+      "1.0.dev456",
+      "1.0a1",
+      "1.0a2.dev456",
+      "1.0a12.dev456",
+      "1.0a12",
+      "1.0b1.dev456",
+      "1.0b2",
+      "1.0b2.post345.dev456",
+      "1.0b2.post345",
+      "1.0b2-346",
+      "1.0c1.dev456",
+      "1.0c1",
+      "1.0rc2",
+      "1.0c3",
+      "1.0",
+      "1.0.post456.dev34",
+      "1.0.post456",
+      "1.1.dev1",
+      "1.2+123abc",
+      "1.2+123abc456",
+      "1.2+abc",
+      "1.2+abc123",
+      "1.2+abc123def",
+      "1.2+1234.abc",
+      "1.2+123456",
+      "1.2.r32+123456",
+      "1.2.rev33+123456",
+      "1!1.0b2.post345.dev456",
+      "1!1.0",
+      "1!1.0.post456.dev34",
+      "1!1.0.post456",
+      "1!1.2.rev33+123456",
+      "2!2.3.4.alpha5.rev6.dev7+abc89"
+    ]
+
+    (0..(versions.length - 1)).each do |lower_version_index|
+      lower_version = described_class.parse(versions.at(lower_version_index))
+      
+      it "#{lower_version} should be equal to #{lower_version}" do
+        expect(lower_version == lower_version).to eq(true)
+      end
+
+      ((lower_version_index + 1)..(versions.length - 1)).each do |greater_version_index|
+        greater_version = described_class.parse(versions.at(greater_version_index))
+
+        it "#{lower_version} should not be equal to #{greater_version}" do
+          expect(lower_version != greater_version).to eq(true)
+        end
+
+        it "#{lower_version} should be lower than #{greater_version}" do
+          expect(lower_version < greater_version).to eq(true)
+        end
+
+        it "#{greater_version} should be greater than #{lower_version}" do
+          expect(greater_version > lower_version).to eq(true)
+        end
+      end
```

Comment:
you can use Array#combination to make testing pairs.

https://apidock.com/ruby/Array/combination